engines/awe/sound.cpp

```diff
@@ -0,0 +1,614 @@
+	void mixWav(int16 *samples, int count) {
+		for (int i = 0; i < count; i += 2) {
+			uint32 pos = _pos.getInt();
+			_pos.offset += _pos.inc;
+			if (pos >= _len) {
+				if (_loopLen != 0) {
+					pos = 0;
+					_pos.offset = _pos.inc;
+				} else {
+					_data = 0;
+					break;
+				}
+			}
+			if (stereo) {
+				pos *= 2;
+			}
+			int valueL;
+			if (bits == 8) { // U8
+				valueL = toS16(_data[pos]) * _volume / 64;
+			} else { // S16
+				valueL = ((int16)READ_LE_UINT16(&_data[pos * sizeof(int16)])) * _volume / 64;
+			}
+			*samples = mixS16(*samples, valueL);
+			++samples;
+
+			int valueR;
+			if (!stereo) {
+				valueR = valueL;
+			} else {
+				if (bits == 8) { // U8
+					valueR = toS16(_data[pos + 1]) * _volume / 64;
+				} else { // S16
+					valueR = ((int16)READ_LE_UINT16(&_data[(pos + 1) * sizeof(int16)])) * _volume / 64;
+				}
+			}
+			*samples = mixS16(*samples, valueR);
+			++samples;
+		}
+	}
+};
+
+static const byte *loadWav(const byte *data, int &freq, int &len, bool &bits16, bool &stereo) {
+	uint32 riffMagic = READ_LE_UINT32(data);
+	if (riffMagic != TAG_RIFF) return 0;
+	uint32 riffLength = READ_LE_UINT32(data + 4);
+	uint32 waveMagic = READ_LE_UINT32(data + 8);
+	if (waveMagic != TAG_WAVE) return 0;
+	uint32 offset = 12;
+	uint32 chunkMagic, chunkLength = 0;
+	// find fmt chunk
+	do {
+		offset += chunkLength + (chunkLength & 1);
+		if (offset >= riffLength) return 0;
+		chunkMagic = READ_LE_UINT32(data + offset);
+		chunkLength = READ_LE_UINT32(data + offset + 4);
+		offset += 8;
+	} while (chunkMagic != TAG_fmt);
+
+	if (chunkLength < 14) return 0;
+	if (offset + chunkLength >= riffLength) return 0;
+
+	// read format
+	int formatTag = READ_LE_UINT16(data + offset);
+	int channels = READ_LE_UINT16(data + offset + 2);
+	int samplesPerSec = READ_LE_UINT32(data + offset + 4);
+	int bitsPerSample = 0;
+	if (chunkLength >= 16) {
+		bitsPerSample = READ_LE_UINT16(data + offset + 14);
+	} else if (formatTag == 1 && channels != 0) {
+		int blockAlign = READ_LE_UINT16(data + offset + 12);
+		bitsPerSample = (blockAlign * 8) / channels;
+	}
+	// check supported format
+	if ((formatTag != 1) || // PCM
+		(channels != 1 && channels != 2) || // mono or stereo
+		(bitsPerSample != 8 && bitsPerSample != 16)) { // 8bit or 16bit
+		warning("Unsupported wave file");
+		return 0;
+	}
+
+	// find data chunk
+	do {
+		offset += chunkLength + (chunkLength & 1);
+		if (offset >= riffLength) return 0;
+		chunkMagic = READ_LE_UINT32(data + offset);
+		chunkLength = READ_LE_UINT32(data + offset + 4);
+		offset += 8;
+	} while (chunkMagic != TAG_data);
+
+	uint32 lengthSamples = chunkLength;
+	if (offset + lengthSamples - 4 > riffLength) {
+		lengthSamples = riffLength + 4 - offset;
+	}
+	if (channels == 2) lengthSamples >>= 1;
+	if (bitsPerSample == 16) lengthSamples >>= 1;
+
+	freq = samplesPerSec;
+	len = lengthSamples;
+	bits16 = (bitsPerSample == 16);
+	stereo = (channels == 2);
+
+	return data + offset;
+}
+
+struct Mixer_impl {
```

Comment:
This looks awfully similar to our own mixer implementation. Perhaps, nothing of the below is needed?

Reply:
All this block is hidden in a "#ifdef todo" block and is currently only kept to help future implementation. Music and most of the sound is STUBed for the moment.